zero/src/rpc/jumpdest.rs

```diff
@@ -0,0 +1,516 @@
+/// to simulating the JumpDest analysis.
+const TIMEOUT_LIMIT: Duration = Duration::from_secs(2 * 60);
+
+/// Structure of Etheruem memory
+type Word = [u8; 32];
+const WORDSIZE: usize = std::mem::size_of::<Word>();
+
+/// Pass `true` for the components needed.
+fn structlog_tracing_options(stack: bool, memory: bool, storage: bool) -> GethDebugTracingOptions {
+    GethDebugTracingOptions {
+        config: GethDefaultTracingOptions {
+            disable_stack: Some(!stack),
+            // needed for CREATE2
+            disable_memory: Some(!memory),
+            disable_storage: Some(!storage),
+            ..GethDefaultTracingOptions::default()
+        },
+        tracer: None,
+        ..GethDebugTracingOptions::default()
+    }
+}
+
+/// Predicate that determines whether a `StructLog` that includes memory is
+/// required.
+fn trace_contains_create(structlog: &[StructLog]) -> bool {
+    structlog
+        .iter()
+        .any(|entry| entry.op == "CREATE" || entry.op == "CREATE2")
+}
+
+/// Gets the lightest possible structlog for transcation `tx_hash`.
+pub(crate) async fn get_normalized_structlog<ProviderT, TransportT>(
+    provider: &ProviderT,
+    tx_hash: &B256,
+) -> Result<Option<Vec<StructLog>>, RpcError<TransportErrorKind>>
+where
+    ProviderT: Provider<TransportT>,
+    TransportT: Transport + Clone,
+{
+    let inner = async {
+        // Optimization: It may be a better default to pull the stack immediately.
+        let stackonly_structlog_trace = provider
+            .debug_trace_transaction(*tx_hash, structlog_tracing_options(true, false, false))
+            .await?;
+
+        let stackonly_structlog_opt: Option<Vec<StructLog>> =
+            normalize_structlog(&stackonly_structlog_trace).await;
+
+        let need_memory = stackonly_structlog_opt
+            .as_deref()
+            .is_some_and(trace_contains_create);
+        trace!("Need structlog with memory: {need_memory}");
+
+        if need_memory.not() {
+            return Ok(stackonly_structlog_opt);
+        };
+
+        let memory_structlog_fut = provider.debug_trace_transaction(
+            *tx_hash,
+            structlog_tracing_options(true, need_memory, false),
+        );
+
+        let memory_structlog = normalize_structlog(&memory_structlog_fut.await?).await;
+
+        Ok::<Option<Vec<_>>, RpcError<TransportErrorKind>>(memory_structlog)
+    };
+
+    match timeout(TIMEOUT_LIMIT, inner).await {
+        Err(ellapsed_error) => Err(RpcError::Transport(TransportErrorKind::Custom(Box::new(
+            ellapsed_error,
+        )))),
+        Ok(structlog_res) => Ok(structlog_res?),
+    }
+}
+
+/// Provides a way to check in constant time if an address points to a
+/// precompile.
+fn precompiles() -> &'static HashSet<Address> {
+    static PRECOMPILES: OnceLock<HashSet<Address>> = OnceLock::new();
+    PRECOMPILES.get_or_init(|| {
+        HashSet::<Address>::from_iter((1..=0xa).map(|x| Address::from(U160::from(x))))
+    })
+}
```

Comment:
This is network specific, i.e. see `trace_decoder/src/core.rs` related code

Reply:
```bash
rg "precompile" -i trace_decoder
trace_decoder/src/core.rs
499:                    const PRECOMPILE_ADDRESSES: Range<alloy::primitives::Address> =
503:                    if receipt.status || !PRECOMPILE_ADDRESSES.contains(&addr.compat()) {
```

I don't see anything network specific here, but we could at least share the code.

Reply:
Done sharing.